### PR TITLE
helpers/toLuaObject: add support for raw keys

### PR DESCRIPTION
**Pull request by GaetanLepage**
This allows to have "raw lua" for table keys.

This:
```nix
{
  __raw__foo = 1;
  bar = 2;
}
```

Would translate into
```lua
{
  [foo] = 1;
  ['bar'] = 2;
}
```

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Hmm, what do you think about the prefix ? Maybe `__rawKey__` would be better no ?

File: lib/utils.nix [hunk 1]
```diff
@@ -11,6 +11,9 @@ with lib;
     "__empty" = null;
   };
 
+  # Turns all the keys of an attrs to raw lua.
+  rawKeysAttrs = mapAttrs' (n: v: nameValuePair "__rawKey__${n}" v);
```
Comment (GaetanLepage) [inline lib/utils.nix:15]:
```suggestion
  toRawKeysAttrs = mapAttrs' (n: v: nameValuePair "__rawKey__${n}" v);
```
?

Comment (MattSturgeon) [reply lib/utils.nix:15]: Agreed. I'm unsure on `toRawKeysAttrs` vs `toRawKeyAttr` though

Comment (GaetanLepage) [reply lib/utils.nix:15]: Hmm, maybe `toRawKeys` ? X) I can't choose.

**Comment by MattSturgeon** [comment]
So to set a raw type value with a raw key, you'd use this?

```nix
{
  __rawKey__someLuaHere.__raw = "moreLuaHere";
}
```

I agree we should support raw keys. Prefixes feel a little messy though.

Would something along these lines work (better/worse)?

```nix
{
  __key.__raw = "someLua";
  __value = "a string";
}
```

**Comment by GaetanLepage** [comment]
> So to set a raw type value with a raw key, you'd use this?
> 
> ```nix
> {
>   __rawKey__someLuaHere.__raw = "moreLuaHere";
> }
> ```
> 
> I agree we should support raw keys. Prefixes feel a little messy though.
> 
> Would something along these lines work (better/worse)?
> 
> ```nix
> # a __key __value pair:
> {
>   __key.__raw = "someLua";
>   __value = "a string";
> }
> ```

I am not sure to get the proposition.
What is important is to support any combination of raw/normal key and raw/normal value.

```nix
{
  normal_key_1 = "normal value";
  normal_key_2.__raw = "raw lua value";
  __raw__some_raw_key_1 = "normal value";
  __raw__some_raw_key_2.__raw = "raw lua value";
}
```
Apart from prefixes, I don't see any option (but I could miss some !).
Important note: raw keys will most likely never be popular. Indeed, it is the first time that we encounter a need for them.
So, the usability is not crucial for users.

**Comment by MattSturgeon** [comment]
> > Would something along these lines work (better/worse)?
> > ```nix
> > # a __key __value pair:
> > {
> >   __key.__raw = "someLua";
> >   __value = "a string";
> > }
> > ```
> 
> I am not sure I get the proposition. [...] Apart from prefixes, I don't see any option (but I could miss some !).

My "pair" proposal was intended to allow for slightly easier to read (if more verbose) definitions.

In hindsight it'd be a little clunky as it'd have to be mapped from a list of pairs (`[ {pair} ]`) into a lua table (string).

_I guess_ that could be an alternative way to define mixed list/dict tables though (a list of normal types alongside pairs).

As with any `[pair]->{dict}` mapping, you'd have key duplication issues too.

```nix
luaTableOption = [
  {
    __key.__raw = "luaKey";
    __value = "string";
  }
  {
    __key = "normal_key";
    __value.__raw = "lua code";
  }
  {
    __key = "another_key";
    __value = "string";
  }
  "un-keyed table entry"
  {
    __raw = "print('un-keyed table entry with raw lua')";
  }
];
```

> What is important is to support any combination of raw/normal key and raw/normal value.

Agreed

> Important note: raw keys will most likely never be popular. Indeed, it is the first time that we encounter a need for them. So, the usability is not crucial for users.

Agreed, it'll likely only be used internally, but some users may find it useful too.

---

I think my `__rawKey__` prefix legibility concerns would be addressed better by the `toRawKeyAttr` function you've already added, and/or possibly a `mkRawKey` style function.

Consider:

```nix
{
  someOption =
    { /* normal attrs */ }
    // (helpers.mkRawKey "someLuaKey" "some value");
  # vs
  otherOption =
    { /* normal attrs */ }
    // (helpers.toRawKeyAttrs {
      "someLuaKey" = "some value";
    });
  # vs
  lowLevelOption = {
    /* normal attrs */
    "__rawKey__someLuaKey" = "some value";
  };
```

File: lib/utils.nix (hunk 1 above)
Comment (MattSturgeon) [reply lib/utils.nix:15]: I think either `toRawKeys`, `toRawKeyAttrs` (one `s` total), or `attrsToRawKeys`.

We could also have a companion function `mkRawKey = key: value: toRawKey { ${name} = value; }`

File: lib/utils.nix [hunk 2]
```diff
@@ -11,6 +11,12 @@ with lib;
     "__empty" = null;
   };
 
+  # Turn all the keys of an attrs to raw lua.
+  toRawKey = mapAttrs' (n: v: nameValuePair "__rawKey__${n}" v);
```
Comment (MattSturgeon) [inline lib/utils.nix:15]: I think this should be a plural:

```suggestion
  toRawKeys = mapAttrs' (n: v: nameValuePair "__rawKey__${n}" v);
```

(also update usage in `mkRawKey`, of course)

File: lib/utils.nix [hunk 3]
```diff
@@ -11,6 +11,12 @@ with lib;
...
+
+  # Create a 1-element attrs with a raw lua key
```
Comment (MattSturgeon) [inline lib/utils.nix:17]:
```nix
  /**
    Create a 1-element attrs with a raw lua key.

    # Example

    ```nix
    mkRawKey "foo" 1
    => { __rawKey__foo = 1; }
    ```

    # Type

    ```
    mkRawKey :: String -> String -> AttrSet
    ```

    # Arguments

    - [n] The attribute name (raw lua)
    - [v] The attribute value
  */
```

File: lib/utils.nix [hunk 4]
```diff
@@ -11,6 +11,12 @@ with lib;
     "__empty" = null;
   };
 
+  # Turn all the keys of an attrs to raw lua.
```
Comment (MattSturgeon) [inline lib/utils.nix:14]: Use [RFC145](https://github.com/nixos/rfcs/blob/master/rfcs/0145-doc-strings.md#migration-example)?

```nix
  /**
    Turn all the keys of an attrs into raw lua.

    # Example

    ```nix
    toRawKeys { foo = 1; bar = 2; }
    => { __rawKey__foo = 1; __rawKey__bar = 2; }
    ```

    # Type

    ```
    toRawKeys :: AttrSet -> AttrSet
    ```
  */
```

File: lib/to-lua.nix [hunk 5]
```diff
@@ -15,12 +15,17 @@ rec {
         + (concatStringsSep "," (
           mapAttrsToList (
             n: v:
-            if (builtins.match "__unkeyed.*" n) != null then
-              toLuaObject v
+            let
+              valueString = toLuaObject v;
```
Comment (MattSturgeon) [inline lib/to-lua.nix:19]:
```suggestion
              lua = toLuaObject v;
```
What do you think?

**Review by MattSturgeon — COMMENTED** [review]
Final suggestions, use or ignore as you wish.

Also consider whether we should add anything to the user-guide.

File: lib/utils.nix (hunk 4 above)
Comment (GaetanLepage) [reply lib/utils.nix]: Done

File: lib/to-lua.nix (hunk 5 above)
Comment (GaetanLepage) [reply lib/to-lua.nix:19]: I think that `valueString` is clearer, but again, no strong opinion X)

Comment (MattSturgeon) [reply lib/to-lua.nix:19]: I think `lua` implies the string has already been rendered/formatted as lua, but I also don't have a strong opinion.

`luaValue` and `luaString` could also work, though `lua` has the advantage of also being less verbose.

**Comment by GaetanLepage** [comment]
Sure, we have to wait for traxys' ACK before merging?

**Comment by MattSturgeon** [comment]
> Sure, we have to wait for traxys' ACK before merging.

Unless you think that's overkill

**Comment by GaetanLepage** [comment]
@traxys can you have a quick look at this ?
I would like to get the barbar refactoring merged before 24.05.

**Pull request merged**